query.go

```diff
@@ -85,15 +109,34 @@ func DoneChan(doneChan chan<- struct{}) QueryOption {
 	}
 }
 
-type spMsg struct {
-	sp  *ServerPeer
-	msg wire.Message
-}
+// checkPeer alows the caller to query a single peer. Should be called in its
+// own goroutine. Exits when quit channel is closed or a message on it is
+// received.
+func checkPeer(sp *ServerPeer, subscription spMsgSubscription,
+	queryMsg wire.Message, quit <-chan struct{}) {
+
+	// Don't do anything if the peer isn't connected.
+	if sp == nil || !sp.Connected() {
+		return
+	}
+
+	// Subscribe to the peer's received messages.
+	sp.subscribeRecvMsg(subscription)
+	defer sp.unsubscribeRecvMsgs(subscription)
+
+	// Send the query to the peer. Exit if we quit before reading the sent
+	// channel.
+	sentChan := make(chan struct{}, 1)
+	sp.QueueMessageWithEncoding(queryMsg, sentChan, wire.WitnessEncoding)
+	select {
+	case <-sentChan:
+	case <-quit:
+	}
 
-type spMsgSubscription struct {
-	msgChan  chan<- spMsg
-	quitChan <-chan struct{}
-	wg       *sync.WaitGroup
+	// Wait for a quit signal.
```

Comment:
What's the use of the select if it falls through anyway? Perhaps should also pass in the quit channel of the external server if it's available?

Reply:
Ah, thanks, that was just a leftover artifact. I had it passing in the external quit channels originally but I had 2 more already (used differently by different schedulers) and it seems like it'd make more sense to make the caller responsible for a single quit channel if we are going to factor out the scheduler into a functional option. If I could select on a slice of channels, that'd be another option.

Reply:
I just looked at the code again. The reason that second quit is there is to delay the unsubscribe. But since we have the timeout in the calling function which sends over/closes that channel, we can just subscribe and unsubscribe and send query messages right in the same function, without starting a new goroutine. The change seems to test well, and now we've eliminated yet another goroutine.